crossbar/worker/proxy.py

```diff
@@ -0,0 +1,537 @@
+                "endpoint": {
+                    "type": "tcp",
+                    "port": 9999,
+                    "interface": "localhost"
+                },
+                "auth": {
+                    "proxy": {}
+                }
+            }
+        ]
+    """
+    log = txaio.make_logger()
+    name = u'proxy'
+
+    def __init__(self, session, config):
+        self._realm = session._realm
+        self._authid = session._authid
+        self._authrole = session._authrole
+        self._authmethod = session._authmethod
+        self._authextra = session._authextra
+        self._authprovider = 'proxy'
+
+    def hello(self, realm, details):
+        self.log.info(
+            "PendingAuthProxy: hello {realm} {details}",
+            realm=realm,
+            details=details,
+        )
+        self._realm = realm
+        self._authid = details.authid
+        self._authrole = details.authrole
+        self._authextra = details.authextra
+        return self._accept()
+
+
+class AuthProxy(PendingAuth):
+    name = u'proxy'
+
+    def __init__(self, **kw):
+        self._realm = kw.get('authrealm', None)
+        self._authid = kw.get('authid', 'anonymous')
+        self._authrole = kw.get('authrole', None)
+        self._args = kw
+
+    @property
+    def authextra(self):
+        return self._args.get(u'authextra', dict())
+
+    def on_challenge(self, session, challenge):
+        raise RuntimeError("on_challenge called on proxy authentication")
+
+    def on_welcome(self, msg, authextra):
+        self.log.info(
+            "AuthProxy.on_welcome: {message} {extra}",
+            message=msg,
+            extra=authextra,
+        )
+        return {
+            "authextra": {
+                "authid": self._authid,
+                "realm": self._realm,
+            }
+        }
+
+
+auth.IAuthenticator.register(AuthProxy)
+
+
+class ProxyServerSession(RouterSession):
+    """
+    The autobahn.wamp.interfaces.ITransportHandler for Proxies. This
+    forwards every incoming message to the other side's transport
+    (with no further processing) with the *exception* of the messages
+    in .do_not_forward
+    """
+    _backend = None
+    log = txaio.make_logger()
+
+    # types of messages we do not forward but pass up to the parent
+    # class for normal-style Router processing
+    do_not_forward = (
+        message.Hello,
+        message.Welcome,
+        message.Authenticate,
+    )
+
+    def set_backend(self, other_session):
+        self._backend = other_session
+
+    def onMessage(self, msg):
+        if isinstance(msg, self.do_not_forward):
+            return super(ProxyServerSession, self).onMessage(msg)
+        # TODO: should always destroy 'the other side' if one side
+        # (including the backend) goes away (e.g. message IDs would
+        # get out of sync if we just auto re-connected to the backend)
+        if not self._backend or not self._backend._transport:
```

Comment:
I *think* this means we need to watch for Goodbye etc messages and take appropriate action..?

Reply:
oh, I see: proxy is forwarding messages at the transport layer without actually creating a WAMP session to the backend using our machinery in eg ApplicationSession. 

so with the current design, yes, we would need to add all WAMP session establishment/close message handling to proxy. plus auth handling as this is part of the existing machinery.

that is, proxy needs to process these messages

* 531:class Hello(Message):
* 776:class Welcome(Message):
* 1027:class Abort(Message):
* 1113:class Challenge(Message):
* 1188:class Authenticate(Message):
* 1263:class Goodbye(Message):

and proxy can plain forward (both directions) all other messages - parsed/reserialized under the hood because the backend connection can use a different transport+serializer than the incoming client connection (IOW: we cannot simply forward at the websocket level)

see also my comment at the very end of this PR